### PR TITLE
refactor: introduce Repository context

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: mergify_engine/context.py [hunk 1]
```diff
@@ -63,25 +63,74 @@ class Installation:
     subscription: subscription.Subscription
     client: github.GithubInstallationClient
 
+    repositories: "typing.Dict[github_types.GitHubRepositoryName, Repository]" = (
+        dataclasses.field(default_factory=dict)
+    )
```
Comment (jd) [inline mergify_engine/context.py:68]: Do you really need to store them?

File: mergify_engine/context.py [hunk 2]
```diff
@@ -63,25 +63,74 @@ class Installation:
...
+    installation: Installation
+    name: str
+    pull_contexts: "typing.Dict[github_types.GitHubPullRequestNumber, Context]" = (
+        dataclasses.field(default_factory=dict)
+    )
+
+    @property
+    def base_url(self) -> str:
+        """The URL prefix to make GitHub request."""
+        return f"/repos/{self.installation.owner_login}/{self.name}"
```
Comment (jd) [inline mergify_engine/context.py:94]: This is actually `github_types.GitHubRepository['url']` so I'm wondering if we really need this.
If it's possible to use the GitHubRepository URL that makes it easier to support different URL scheme etc :)

File: mergify_engine/context.py [hunk 3]
```diff
@@ -63,25 +63,74 @@ class Installation:
...
+        return self.repositories[name]
+
+
+@dataclasses.dataclass
+class Repository(object):
+    installation: Installation
+    name: str
+    pull_contexts: "typing.Dict[github_types.GitHubPullRequestNumber, Context]" = (
+        dataclasses.field(default_factory=dict)
+    )
```
Comment (jd) [inline mergify_engine/context.py:89]: Do we need to store this either?

File: mergify_engine/context.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/context.py:68]: This is planned to be replace this Context construction here: https://github.com/Mergifyio/mergify-engine/pull/1729/files#diff-e709bf7f7a0ceb20e0cfafd3b8ab1e163a0c04e84262742607dc6828cf868770R258

File: mergify_engine/context.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/context.py:89]: This is planned to be replace this Context construction here: https://github.com/Mergifyio/mergify-engine/pull/1729/files#diff-e709bf7f7a0ceb20e0cfafd3b8ab1e163a0c04e84262742607dc6828cf868770R258

File: mergify_engine/context.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/context.py:95]: This url is currently not forwarded from the webhook to the worker.

**Comment by sileht** [comment]
@jd ?

**Review by jd — APPROVED** [review]

**Pull request merged**